src/sync.rs

```diff
@@ -0,0 +1,302 @@
+            assert_eq!(thread, current_owner, "mutex already locked by another thread");
+            assert!(
+                mutex.lock_count > 0,
+                "invariant violation: lock_count == 0 iff the thread is unlocked"
+            );
+        } else {
+            mutex.owner = Some(thread);
+        }
+        mutex.lock_count = mutex.lock_count.checked_add(1).unwrap();
+    }
+
+    /// Unlock by decreasing the lock count. If the lock count reaches 0, unset
+    /// the owner.
+    fn mutex_unlock(&mut self, id: MutexId) -> Option<(ThreadId, usize)> {
+        let this = self.eval_context_mut();
+        let mutex = &mut this.machine.threads.sync.mutexes[id];
+        if let Some(current_owner) = mutex.owner {
+            mutex.lock_count = mutex
+                .lock_count
+                .checked_sub(1)
+                .expect("invariant violation: lock_count == 0 iff the thread is unlocked");
+            if mutex.lock_count == 0 {
+                mutex.owner = None;
+            }
+            Some((current_owner, mutex.lock_count))
+        } else {
+            None
+        }
+    }
+
+    #[inline]
+    /// Take a thread out the queue waiting for the lock.
+    fn mutex_enqueue(&mut self, id: MutexId, thread: ThreadId) {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.mutexes[id].queue.push_back(thread);
+    }
+
+    #[inline]
+    /// Take a thread out the queue waiting for the lock.
+    fn mutex_dequeue(&mut self, id: MutexId) -> Option<ThreadId> {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.mutexes[id].queue.pop_front()
+    }
+
+    #[inline]
+    /// Create state for a new read write lock.
+    fn rwlock_create(&mut self) -> RwLockId {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.rwlocks.push(Default::default())
+    }
+
+    #[inline]
+    /// Check if locked.
+    fn rwlock_is_locked(&mut self, id: RwLockId) -> bool {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.rwlocks[id].writer.is_some()
+            || !this.machine.threads.sync.rwlocks[id].readers.is_empty()
+    }
+
+    #[inline]
+    /// Check if write locked.
+    fn rwlock_is_write_locked(&mut self, id: RwLockId) -> bool {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.rwlocks[id].writer.is_some()
+    }
+
+    /// Add a reader that collectively with other readers owns the lock.
+    fn rwlock_reader_add(&mut self, id: RwLockId, reader: ThreadId) {
+        let this = self.eval_context_mut();
+        assert!(!this.rwlock_is_write_locked(id), "the lock is write locked");
+        let count = this.machine.threads.sync.rwlocks[id].readers.entry(reader).or_insert(0);
+        *count += 1;
+    }
+
+    /// Try removing the reader. Returns `true` if succeeded.
+    fn rwlock_reader_remove(&mut self, id: RwLockId, reader: ThreadId) -> bool {
```

Comment:
Is "remove" here "unlock"? If so, I think we need better terminology... "remove" is too similar to "dequeue".

Either "unlock" or "release" would be appropriate, IMO.